vermouth/processors/molecule_grouper.py

```diff
@@ -0,0 +1,295 @@
+                            weight=round(precision*violation_cost**2),
+                            capacity=tolerances[clust_n_idx-len(data_nodes)])
+        flow_graph.add_edge(artificial_node, clust_n_idx,
+                            weight=round(precision*violation_cost**2),
+                            capacity=tolerances[clust_n_idx-len(data_nodes)])
+        # Demand for cluster nodes
+        flow_graph.nodes[clust_n_idx]['demand'] = clust_sizes[clust_n_idx - len(data)]
+    for data_n_idx in data_nodes:
+        # Data supply
+        flow_graph.nodes[data_n_idx]['demand'] = -1
+    for data_n_idx, clust_n_idx in itertools.product(data_nodes, clust_nodes):
+        # Edges between data and clusters
+        # Capacity is meaningless, since data nodes have only 1 supply to give,
+        # but it seems to help the network simplex a little.
+        flow_graph.add_edge(data_n_idx, clust_n_idx, capacity=1)
+
+    # Start clustering iterations.
+    iter = 0
+    flow_dict = {}
+    while iter < max_iter:
+        # Step 1: set up edge weights based on clustroid positions
+        # TODO: PBC
+        dists = np.sum((data[:, np.newaxis] - clusters[np.newaxis, :])**2, axis=-1)
+        # When clusters are empty their clustroids become nan, which we can't
+        # round to an int.
+        dists[np.isnan(dists)] = 10**2
+        for (data_idx, clust_idx), dist in np.ndenumerate(dists):
+            # Network simplex (and capacity scaling) does not work with float
+            # weights according to nx documentation. So round the distances to
+            # an appropriate int. Magic.
+            flow_graph.edges[data_idx, clust_idx + len(data)]['weight'] = round(precision*dist)
+
+        prev_weights = flow_dict
+        cost, flow_dict = simplex_method(flow_graph)
+
+        if prev_weights == flow_dict:
+            break
+        # Step 2: make memberships and calculate new clustroid positions
+        memberships = np.zeros((len(data), num_clusters))
+        for data_idx in data_nodes:
+            flow = flow_dict[data_idx]
+            memberships[data_idx] = [flow[clust_n_idx] for clust_n_idx in clust_nodes]
+
+        # TODO: PBC
+        weights = memberships / memberships.sum(axis=0, keepdims=True)
+        clusters = weights.T.dot(data)
+
+        iter += 1
+
+    return cost/precision, clusters, memberships, iter
+
+
+def group_molecules(system, selector, size_tries=10, **kwargs):
+    """
+    Clusters molecules in `system` into groups of 4 \u00b1 1. Only molecules
+    selected with `selector` will be taken.
+
+    Parameters
+    ----------
+    system
+    selector
+    size_tries: int
+        The number of clusters to try.
+    **kwargs
+        Passed on to :func:`constrained_kmeans`.
+
+    Returns
+    -------
+    None
+        `system` is modified in place.
+    """
+    water_mols = [(mol_idx, mol) for (mol_idx, mol) in enumerate(system.molecules) if selector(mol)]
```

Comment:
As a more general note, I think in the future it would be useful to have some correspondence between a molecule and it's index in the topology file to specifically target certain molecules. So far much of the vermouth infrastructure assumes you wanna do something on a single molecule or selection based on name, res type etc. Perhaps this could become an attribute of system.

Reply:
I don't quite understand what you mean here